fl4health/model_bases/feature_extractor_buffer.py

```diff
@@ -0,0 +1,177 @@
+
+        self.accumulate_features: bool = False
+        self.extracted_features_buffers: Dict[str, List[torch.Tensor]] = {
+            layer: [] for layer in flatten_feature_extraction_layers.keys()
+        }
+
+    def enable_accumulating_features(self) -> None:
+        """
+        Enables the accumulation of features in the buffers for multiple forward passes.
+
+        This method sets the `accumulate_features` flag to True, allowing the model to accumulate features
+        in the buffers for multiple forward passes. This can be useful in scenarios where you want to extract
+        features from intermediate layers of the model during inference.
+        """
+        self.accumulate_features = True
+
+    def disable_accumulating_features(self) -> None:
+        """
+        Disables the accumulation of features in the buffers.
+
+        This method sets the `accumulate_features` attribute to False, which prevents the buffers from accumulating
+        features and overwrits them for each forward pass.
+        """
+        self.accumulate_features = False
+
+    def clear_buffers(self) -> None:
+        """
+        Clears the extracted features buffers for all layers.
+        """
+        self.extracted_features_buffers = {layer: [] for layer in self.flatten_feature_extraction_layers.keys()}
+
+    def get_hierarchical_attr(self, module: nn.Module, layer_hierarchy: List[str]) -> nn.Module:
+        """
+        Traverse the hierarchical attributes of the module to get the desired attribute, as hooks should be
+        registered to specific layers of the model, not to nn.Sequential or nn.ModuleList.
+        """
+        if len(layer_hierarchy) == 1:
+            return getattr(module, layer_hierarchy[0])
+        else:
+            return self.get_hierarchical_attr(getattr(module, layer_hierarchy[0]), layer_hierarchy[1:])
+
+    def find_last_common_prefix(self, prefix: str, layers_name: List[str]) -> str:
+        """
+        Check the model's list of named modules to filter any layer that starts with the given prefix and
+        return the last one. Here we assume the list of named modules is sorted in the order of the model's
+        forward pass with depth-first traversal. This will allow the user to specify the generic name of the
+        layer instead of the full hierarchical name.
+        """
+        filtered_layers = [layer for layer in layers_name if layer.startswith(prefix)]
+
+        # Return the last element that matches the criteria
+        return filtered_layers[-1]
+
+    def _maybe_register_hooks(self) -> None:
+        """
+        Checks if hooks are already registered and registers them if not.
+        Hooks extract the intermediate feature as output of the selected layers in the model.
+        """
+        if len(self.fhooks) == 0:
+            log(INFO, "Starting to register hooks:")
+            named_layers = list(dict(self.model.named_modules()).keys())
```

Comment:
is the conversion to dict then to list necessary or can it just be converted to a list directly?

Reply:
I think we need the dict as return of .named_modules() returns an iterator over all modules in the network. However maybe we can omit the list.

Reply:
Looks like I can omit dict conversion,

Reply:
.named_modules() returns an iterator over all modules in the network. If I omit dict I get 'generator' object has no attribute 'keys' error. And if I omit list the type of .keys() is dict_keys[Any, Any] which raises error. So I think we need both.

Reply:
That makes sense. I thought `named_modules` returned a dictionary.